README.md
# Autosub

<escape><a href="https://travis-ci.org/BingLingGroup/autosub"><img src="https://travis-ci.org/BingLingGroup/autosub.svg?branch=alpha"></img></a> <a href="https://app.fossa.io/projects/git%2Bgithub.com%2FBingLingGroup%2Fautosub"><img src="https://app.fossa.io/api/projects/git%2Bgithub.com%2FBingLingGroup%2Fautosub.svg?type=shield"></img></a></escape>

[简体中文](docs/README.zh-Hans.md)

This repo is not the same as [the original autosub repo](https://github.com/agermanidis/autosub).

This repo has been modified by several people. See the [Changelog](CHANGELOG.md).

<escape><img src="docs/icon/autosub.png" width="128px"></escape>

[autosub icon](docs/icon/autosub.svg) designed by BingLingGroup.

Software: [inkscape](https://inkscape.org/zh/)

Font: [source-han-sans](https://github.com/adobe-fonts/source-han-sans) ([SIL](https://github.com/adobe-fonts/source-han-sans/blob/master/LICENSE.txt))

Color: [Solarized](https://en.wikipedia.org/wiki/Solarized_(color_scheme)#Colors)

### TOC

1. [Description](#description)
2. [License](#license)
3. [Dependencies](#dependencies)
   - 3.1 [Optional Dependencies](#optional-dependencies)
   - 3.2 [Required Dependencies](#required-dependencies)
4. [Download and Installation](#download-and-installation)
   - 4.1 [Branches](#branches)
   - 4.2 [Install on Ubuntu](#install-on-ubuntu)
   - 4.3 [Install on Windows](#install-on-windows)
5. [Workflow](#workflow)
   - 5.1 [Input](#input)
   - 5.2 [Split](#split)
   - 5.3 [Speech-to-Text/Translation API request](#speech-to-texttranslation-api-request)
   - 5.4 [Speech-to-Text/Translation language support](#speech-to-texttranslation-language-support)
   - 5.5 [Output](#Output)
6. [Usage](#usage)
   - 6.1 [Typical usage](#typical-usage)
     - 6.1.1 [Pre-process Audio](#pre-process-audio)
     - 6.1.2 [Detect Regions](#detect-regions)
     - 6.1.3 [Split Audio](#split-audio)
     - 6.1.4 [Transcribe Audio To Subtitles](#transcribe-audio-to-subtitles)
       - 6.1.4.1 [Google Speech V2](#google-speech-v2)
       - 6.1.4.2 [Google Cloud Speech-to-Text](#google-cloud-speech-to-text)
       - 6.1.4.3 [Google speech config](#google-speech-config)
       - 6.1.4.4 [Output API full response](#output-api-full-response)
       - 6.1.4.5 [Xfyun speech config](#xfyun-speech-config)
       - 6.1.4.6 [Baidu speech config](#baidu-speech-config)
     - 6.1.5 [Translate Subtitles](#translate-subtitles)
   - 6.2 [Options](#Options)
   - 6.3 [Internationalization](#internationalization)
7. [FAQ](#FAQ)
   - 7.1 [Other APIs supports](#other-apis-supports)
   - 7.2 [Batch processing](#batch-processing)
   - 7.3 [proxy support](#proxy-support)
   - 7.4 [macOS locale issue](#macos-locale-issue)
   - 7.5 [Accuracy](#accuracy)
8. [Bugs report](#bugs-report)
9. [Build](#build)

Click up arrow to go back to TOC.

### Description

Autosub is an automatic subtitles generating utility. It can detect speech regions automatically by using Auditok, split the audio files according to regions by using ffmpeg, transcribe speech based on several APIs and translate the subtitles' text by using py-googletrans.

The new features mentioned above are only available in the latest alpha branch. Not available on PyPI or the original repo.

### License

This repo has a different license from [the original repo](https://github.com/agermanidis/autosub).

[GPLv2](LICENSE)

[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2FBingLingGroup%2Fautosub.svg?type=large)](https://app.fossa.io/projects/git%2Bgithub.com%2FBingLingGroup%2Fautosub)

### Dependencies

Autosub depends on these third party softwares or Python site-packages. Much appreciation to all of these projects.

#### Optional dependencies

- [ffmpeg](https://ffmpeg.org/)
- [ffprobe](https://ffmpeg.org/ffprobe.html)
- [langcodes](https://github.com/LuminosoInsight/langcodes)
- [ffmpeg-normalize](https://github.com/slhck/ffmpeg-normalize)
- [python-Levenshtein](https://github.com/ztane/python-Levenshtein)(Used by [fuzzywuzzy](https://github.com/seatgeek/fuzzywuzzy))

For windows user:

- [Build Tools for Visual Studio 2019](https://visualstudio.microsoft.com/downloads/)
  - Used by [marisa-trie](https://github.com/pytries/marisa-trie) when installing.
  - [marisa-trie](https://github.com/pytries/marisa-trie) is the dependency of the [langcodes](https://github.com/LuminosoInsight/langcodes).
  - Probable components installation: MSVC v14 VS 2019 C++ build tools, windows 10 SDK.

#### Required dependencies

- [auditok](https://github.com/amsehili/auditok)
- [pysubs2](https://github.com/tkarabela/pysubs2)
- [wcwidth](https://github.com/jquast/wcwidth)
- [requests](https://github.com/psf/requests)
- [fuzzywuzzy](https://github.com/seatgeek/fuzzywuzzy)
- [progressbar2](https://github.com/WoLpH/python-progressbar)
- [websocket-client](https://github.com/websocket-client/websocket-client)
- [py-googletrans](https://github.com/ssut/py-googletrans)

[requirements.txt](requirements.txt).

About how to install these dependencies, see [Download and Installation](#download-and-installation).

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

### Download and Installation

Except the PyPI version, others include non-original codes not from the original repository.

0.4.0 > autosub

- These versions are only compatible with Python 2.7.

0.5.6a >= autosub >= 0.4.0

- These versions are compatible with both Python 2.7 and Python 3. It doesn't matter if you change the Python version in the installation commands below.

autosub >= 0.5.7a

- These versions are only compatible with Python 3.

ffmpeg, ffprobe, ffmpeg-normalize need to be put on one of these places to let the autosub detect and use them. The following codes are in the [constants.py](autosub/constants.py). Priority is determined in order.

1. Set the following environment variables before running the program: `FFMPEG_PATH`, `FFPROBE_PATH` and `FFMPEG_NORMALIZE_PATH`. It will override the ones located at the environment variable `PATH`. This will be helpful if you don't want to use the one in the `PATH`.
2. Add them to the environment variable `PATH`. No need to worry about if using package manager to install such as using pip to install ffmpeg-normalize and using chocolatey to install ffmpeg.
3. Add them to the same directory as the autosub executable.
4. Add them to the current command line working directory.

About the git installation. If you don't want to install git to use pip [VCS](https://pip.pypa.io/en/stable/reference/pip_install/#vcs-support) support to install python package or just confused with git environment variables, you can manually click that clone and download button to download the source code and use pip to install the source code [locally](https://pip.pypa.io/en/stable/reference/pip_install/#description) by input these commands.

```batch
cd the_directory_contains_the_source_code
pip install .
```

Due to the autosub PyPI project is maintained by the original autosub repo's owner, I can't modify it or upload a project with the same name. Perhaps later when this version of autosub becomes stabler, I will rename and duplicate this repo and then upload it to PyPI.

#### Branches

[alpha branch](https://github.com/BingLingGroup/autosub/tree/alpha)

- Include many changes from [the original repo](https://github.com/agermanidis/autosub). Details in [Changelog](CHANGELOG.md). Codes will be updated when an alpha version have been released. It is stabler than the dev branch

[origin branch](https://github.com/BingLingGroup/autosub/tree/origin)

- Include the least changes from [the original repo](https://github.com/agermanidis/autosub) except all new features in the [alpha branch](https://github.com/BingLingGroup/autosub/tree/alpha). The changes in [origin branch](https://github.com/BingLingGroup/autosub/tree/dev) just make sure there's no critical bugs when the program is running on Windows. Currently isn't maintained.

[dev branch](https://github.com/BingLingGroup/autosub/tree/dev)

- The latest codes will be pushed to this branch. If it works fine, it will be merged to alpha branch when new version released.
- Only used to test or pull request. Don't install them unless you know what you are doing.

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

#### Install on Ubuntu

Include dependencies installation commands.

Install from `alpha` branch.(latest autosub alpha release)

```bash
apt install ffmpeg python3 curl git -y
curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
python3 get-pip.py
pip install git+https://github.com/BingLingGroup/autosub.git@alpha ffmpeg-normalize langcodes
```

Install from `dev` branch.(latest autosub dev version)

```bash
apt install ffmpeg python3 curl git -y
curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
python3 get-pip.py
pip install git+https://github.com/BingLingGroup/autosub.git@dev ffmpeg-normalize langcodes
```

Install from `origin` branch.(autosub-0.4.0a)

```bash
apt install ffmpeg python python-pip git -y
pip install git+https://github.com/BingLingGroup/autosub.git@origin
```

Install from PyPI.(autosub-0.3.12)

```bash
apt install ffmpeg python python-pip -y
pip install autosub
```

Recommend using `python3` and `python-pip3` instead of `python` and `python-pip` after autosub-0.4.0.

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

#### Install on Windows

You can just go to the [release page](https://github.com/BingLingGroup/autosub/releases) and download the latest release for Windows. The click-and-run batches are also in the package. You can manually edit by using Notepad++. Or add the executable files' directory to system environment variables so you can use it as a universal command everywhere in the system if permission is Ok.

Tips: `Shift - Right Click` is the keyboard shortcut for opening a Powershell on current directory. To open an exe at current directory, the format is like `.\autosub`.

Or you can just directly open it and input the args manually though I don't recommend doing this due to its less efficiency.

- The one without pyinstaller suffix is compiled by Nuitka. It's faster than the pyinstaller due to its compiling feature different from pyinstaller which just bundles the application.
- ffmpeg and ffmpeg-normalize are also in the package. The original ffmpeg-normalize doesn't have a standalone version. The standalone version of ffmpeg-normalize is built separately. Codes are [here](https://github.com/BingLingGroup/ffmpeg-normalize).
- If there's anything wrong on the both releases, or the package size and any other things are annoying you, you can just use the traditional pip installation method below.

Or install Python environment(if you still don't have one) from choco and then install the package.

Recommend using [chocolatey](https://chocolatey.org) on windows to install the environment and dependencies.

Choco installation command is for cmd.(not Powershell)

```batch
@"%SystemRoot%\System32\WindowsPowerShell\v1.0\powershell.exe" -NoProfile -InputFormat None -ExecutionPolicy Bypass -Command "iex ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))" && SET "PATH=%PATH%;%ALLUSERSPROFILE%\chocolatey\bin"
```

If you don't have [Build Tools for Visual Studio 2019](https://visualstudio.microsoft.com/downloads/), please install autosub without langcodes.

Install from `alpha` branch.(latest autosub alpha release)

```batch
choco install git python curl ffmpeg -y

CHANGELOG.md
## Changelog

[简体中文](docs/CHANGELOG.zh-Hans.md)

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## TOC

- [Unreleased](#unreleased)
  - [Added](#addedunreleased)
  - [Changed](#changedunreleased)
  - [Fixed](#fixedunreleased)
  - [Removed](#removedunreleased)
- [0.5.7-alpha - 2020-05-06](#057-alpha---2020-05-06)
  - [Added](#added057-alpha)
  - [Changed](#changed057-alpha)
  - [Fixed](#fixed057-alpha)
  - [Removed](#removed057-alpha)
- [0.5.6-alpha - 2020-03-20](#056-alpha---2020-03-20)
  - [Added](#added056-alpha)
  - [Changed](#changed056-alpha)
  - [Deprecated](#deprecated056-alpha)
  - [Fixed](#fixed056-alpha)
- [0.5.5-alpha - 2020-03-04](#055-alpha---2020-03-04)
  - [Added](#added055-alpha)
  - [Fixed](#fixed055-alpha)
- [0.5.4-alpha - 2020-01-31](#054-alpha---2020-01-31)
  - [Added](#added054-alpha)
  - [Removed](#removed054-alpha)
  - [Fixed](#fixed054-alpha)
- [0.5.3-alpha - 2019-12-30](#053-alpha---2019-12-30)
  - [Fixed](#fixed053-alpha)
- [0.5.2-alpha - 2019-11-05](#052-alpha---2019-11-05)
  - [Added](#added052-alpha)
  - [Fixed](#fixed052-alpha)
- [0.5.1-alpha - 2019-08-02](#051-alpha---2019-08-02)
  - [Added](#added051-alpha)
  - [Fixed](#fixed051-alpha)
- [0.5.0-alpha - 2019-07-27](#050-alpha---2019-07-27)
  - [Added](#added050-alpha)
  - [Changed](#changed050-alpha)
  - [Fixed](#fixed050-alpha)
- [0.4.1-alpha - 2019-07-11](#041-alpha---2019-07-11)
  - [Added](#added041-alpha)
  - [Changed](#changed041-alpha)
  - [Fixed](#fixed041-alpha)
- [0.4.0-alpha - 2019-02-17](#040-alpha---2019-02-17)
  - [Fixed](#fixed040-alpha)

Click up arrow to go back to TOC.

### [Unreleased]

#### Added(Unreleased)

- Add method convert_wav in cmdline_utils.py.
- Add the manual translation method.
- Add method sub_to_file in cmdline_utils.py.
- Add complex join-events functions.
- Add Auditok options optimization.
- Add send2trash instead of os.remove in some cases.
- Add word check in method man_get_vtt_words_index.
- Add speed limit in trim in join-events.
- Add method split_dst_lf_src_assfile to split bilingual subtitles file's events.
- Add current working directory support in file renaming.

#### Changed(Unreleased)

- Change the default value for `-et` option into 50.
- Change the control flow in method audio_or_video_prcs by using args.output_files to control.

#### Fixed(Unreleased)

- Fix the issue when not input the `-D` option.
- Fix args.audio_split_cmd replacement issue when using option `-ap`. [issue #122](https://github.com/BingLingGroup/autosub/issues/122)
- Fix default value issue with option `--user-agent`. [issue #127](https://github.com/BingLingGroup/autosub/issues/127)
- Fix last word in line not reading in method YTBWebVTT.from_file.
- Fix wrong return value in method list_to_googletrans. [issue #136](https://github.com/BingLingGroup/autosub/issues/136)
- Fix youtube vtt multiple words using one timestamp issue.

### [0.5.7-alpha] - 2020-05-06

#### Added(0.5.7-alpha)

- Add support for Xun Fei Yun Speech-to-Text WebSocket API.
- Add support for Baidu Automatic Speech Recognition API. [issue #68](https://github.com/BingLingGroup/autosub/issues/68)
- Add chars filter for the transcript result in XfyunWebSocketAPI.
- Add delete_chars in method list_to_googletrans.
- Add merge_src_assfile, merge_bilingual_assfile methods.
- Add stop words to split events in merge_src_assfile method.
- Add punctuations split in merge_src_assfile method.
- Add limitation in SplitIntoAudioPiece with an audio length of at least 4 bytes.
- Add support for src language auto-detection when not input `-SRC` language.

#### Changed(0.5.7-alpha)

- Change the replacement condition of the audio_split_cmd only when the user doesn't modify it.
- Change the MAX_REGION_SIZE_LIMIT into 60 seconds.
- Change all text file input decoding into "utf-8".
- Change the default style selection in subtitles translation.
- Change the loglevel in ffmpeg commands into `-loglevel error`.
- Change DEFAULT_MIN_REGION_SIZE to 0.5.
- Change langcodes into the optional dependency.
- Change LICENSE into GPLv2.

#### Fixed(0.5.7-alpha)

- Fix the size count bug when the last line been split in list_to_googletrans.
- Fix delete_chars issue when using `-of full-src`.
- Fix os.remove() PermissionError in api_xfyun.py.
- Fix DEFAULT_AUDIO_PRCS_CMDS and DEFAULT_CHECK_CMD.
- Fix the issue with path checking in dependency finding.
- Fix Baidu's strange error code handling. [issue #114](https://github.com/BingLingGroup/autosub/issues/114)

#### Removed(0.5.7-alpha)

- Remove Python 2.7 support.

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

### [0.5.6-alpha] - 2020-03-20

#### Added(0.5.6-alpha)

- Add extra environment variables check when finding dependencies. [issue #91](https://github.com/BingLingGroup/autosub/issues/91)
- Add arguments parser input when open it without arguments. [issue #92](https://github.com/BingLingGroup/autosub/issues/92)
- Add subtitles processing when not input `-SRC`.

#### Changed(0.5.6-alpha)

- Change option `-sml` into `-nsml`.
- Change defaults for Auditok.

#### Fixed(0.5.6-alpha)

- Fix Google Speech-to-Text API empty result response bug. [issue #89](https://github.com/BingLingGroup/autosub/issues/89)

#### Deprecated(0.5.6-alpha)

- Deprecate Python 2.7 support.

<escape><a href = "#TOC">&nbsp;↑&nbsp;</a></escape>

### [0.5.5-alpha] - 2020-03-04

#### Added(0.5.5-alpha)


autosub/constants.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Defines constants used by autosub.
"""
# Import built-in modules
import os
import re
import sys
import shlex
import locale
import multiprocessing
from pkg_resources import DistributionNotFound

# Import third-party modules
try:
    from google.cloud.speech_v1p1beta1 import RecognitionConfig  # pylint: disable=unused-import
    IS_GOOGLECLOUDCLIENT = True
except DistributionNotFound:
    IS_GOOGLECLOUDCLIENT = False
from send2trash import send2trash

try:
    import langcodes as langcodes_  # pylint: disable=unused-import
except ImportError:
    langcodes_ = None

# Any changes to the path and your own modules

SUPPORTED_LOCALE = {
    "en_US",
    "zh_CN"
}
# Ref: https://www.gnu.org/software/gettext/manual/html_node/Locale-Names.html#Locale-Names

if getattr(sys, 'frozen', False):
    # If the application is run as a bundle, the pyInstaller bootloader
    # extends the sys module by a flag frozen=True and sets the app
    # path into variable executable'.
    APP_PATH = os.path.dirname(sys.executable)
else:
    APP_PATH = os.path.dirname(__file__)

if sys.platform.startswith('win'):
    DEFAULT_ENCODING = "utf-8-sig"
    IS_UNIX = False
    DELETE_PATH = send2trash
else:
    DEFAULT_ENCODING = "utf-8"
    IS_UNIX = True
    DELETE_PATH = os.remove

LOCALE_PATH = os.path.abspath(os.path.join(APP_PATH, "data/locale"))

EXT_LOCALE = os.path.abspath(os.path.join(os.getcwd(), "locale"))
if os.path.isfile(EXT_LOCALE):
    with open(EXT_LOCALE, encoding='utf-8') as in_file:
        LINE = in_file.readline()
        LINE_LIST = LINE.split()
        if LINE_LIST[0] in SUPPORTED_LOCALE:
            CURRENT_LOCALE = LINE_LIST[0]
        else:
            CURRENT_LOCALE = locale.getdefaultlocale()[0]
else:
    CURRENT_LOCALE = locale.getdefaultlocale()[0]

GOOGLE_SPEECH_V2_API_KEY = "AIzaSyBOti4mM-6x9WDnZIjIeyEU21OpBXqWBgw"
GOOGLE_SPEECH_V2_API_URL = \
    "www.google.com/speech-api/v2/recognize?client=chromium&lang={lang}&key={key}"
XFYUN_SPEECH_WEBAPI_URL = "iat-api.xfyun.cn"
BAIDU_ASR_URL = "http://vop.baidu.com/server_api"
BAIDU_PRO_ASR_URL = "http://vop.baidu.com/pro_api"
BAIDU_TOKEN_URL = "http://openapi.baidu.com/oauth/2.0/token"

if multiprocessing.cpu_count() > 3:
    DEFAULT_CONCURRENCY = multiprocessing.cpu_count() >> 1
else:
    DEFAULT_CONCURRENCY = 2

VTT_TIMESTAMP = re.compile(r'\s*((?:\d+:)?\d{2}:\d{2}.\d{3})\s*-->\s*((?:\d+:)?\d{2}:\d{2}.\d{3})')
VTT_WORD_TIMESTAMP = re.compile(r'<(\d{1,2}):(\d{2}):(\d{2})[.,](\d{2,3})>')

DEFAULT_SRC_LANGUAGE = 'en-US'
DEFAULT_ENERGY_THRESHOLD = 50
DEFAULT_MAX_REGION_SIZE = 10.0
DEFAULT_MIN_REGION_SIZE = 0.5
MIN_REGION_SIZE_LIMIT = 0.3
MAX_REGION_SIZE_LIMIT = 60.0
DEFAULT_CONTINUOUS_SILENCE = 0.2
# Maximum speech to text region length in milliseconds
# when using external speech region control

DEFAULT_DST_LANGUAGE = 'en-US'
DEFAULT_SIZE_PER_TRANS = 4000
DEFAULT_SLEEP_SECONDS = 1

DEFAULT_MAX_SIZE_PER_EVENT = 110
DEFAULT_EVENT_DELIMITERS = r"!()*,.:;?[]^_`~"

DEFAULT_SUBTITLES_FORMAT = 'srt'

DEFAULT_MODE_SET = \
    {'regions', 'src', 'full-src', 'dst', 'bilingual', 'dst-lf-src', 'src-lf-dst'}
DEFAULT_SUB_MODE_SET = {'dst', 'bilingual', 'dst-lf-src', 'src-lf-dst'}
DEFAULT_LANG_MODE_SET = {'s', 'src', 'd'}
DEFAULT_AUDIO_PRCS_MODE_SET = {'o', 's', 'y'}

SPEECH_TO_TEXT_LANGUAGE_CODES = {
    'af-za': 'Afrikaans (South Africa)',
    'am-et': 'Amharic (Ethiopia)',
    'ar-ae': 'Arabic (United Arab Emirates)',
    'ar-bh': 'Arabic (Bahrain)',
    'ar-dz': 'Arabic (Algeria)',
    'ar-eg': 'Arabic (Egypt)',
    'ar-il': 'Arabic (Israel)',
    'ar-iq': 'Arabic (Iraq)',
    'ar-jo': 'Arabic (Jordan)',
    'ar-kw': 'Arabic (Kuwait)',
    'ar-lb': 'Arabic (Lebanon)',
    'ar-ma': 'Arabic (Morocco)',
    'ar-om': 'Arabic (Oman)',
    'ar-ps': 'Arabic (State of Palestine)',
    'ar-qa': 'Arabic (Qatar)',
    'ar-sa': 'Arabic (Saudi Arabia)',
    'ar-tn': 'Arabic (Tunisia)',
    'az-az': 'Azerbaijani (Azerbaijan)',
    'bg-bg': 'Bulgarian (Bulgaria)',
    'bn-bd': 'Bengali (Bangladesh)',
    'bn-in': 'Bengali (India)',
    'ca-es': 'Catalan (Spain)',
    'cmn-hans-cn': 'Chinese, Mandarin (Simplified, China)',
    'cmn-hans-hk': 'Chinese, Mandarin (Simplified, Hong Kong)',
    'cmn-hant-tw': 'Chinese, Mandarin (Traditional, Taiwan)',
    'cs-cz': 'Czech (Czech Republic)',
    'da-dk': 'Danish (Denmark)',
    'de-de': 'German (Germany)',
    'el-gr': 'Greek (Greece)',
    'en-au': 'English (Australia)',
    'en-ca': 'English (Canada)',
    'en-gb': 'English (United Kingdom)',
    'en-gh': 'English (Ghana)',
    'en-ie': 'English (Ireland)',
    'en-in': 'English (India)',
    'en-ke': 'English (Kenya)',
    'en-ng': 'English (Nigeria)',
    'en-nz': 'English (New Zealand)',
    'en-ph': 'English (Philippines)',
    'en-sg': 'English (Singapore)',
    'en-tz': 'English (Tanzania)',
    'en-us': 'English (United States)',
    'en-za': 'English (South Africa)',
    'es-ar': 'Spanish (Argentina)',
    'es-bo': 'Spanish (Bolivia)',
    'es-cl': 'Spanish (Chile)',
    'es-co': 'Spanish (Colombia)',
    'es-cr': 'Spanish (Costa Rica)',
    'es-do': 'Spanish (Dominican Republic)',
    'es-ec': 'Spanish (Ecuador)',
    'es-es': 'Spanish (Spain)',
    'es-gt': 'Spanish (Guatemala)',
    'es-hn': 'Spanish (Honduras)',
    'es-mx': 'Spanish (Mexico)',
    'es-ni': 'Spanish (Nicaragua)',
    'es-pa': 'Spanish (Panama)',
    'es-pe': 'Spanish (Peru)',
    'es-pr': 'Spanish (Puerto Rico)',
    'es-py': 'Spanish (Paraguay)',
    'es-sv': 'Spanish (El Salvador)',
    'es-us': 'Spanish (United States)',
    'es-uy': 'Spanish (Uruguay)',
    'es-ve': 'Spanish (Venezuela)',
    'eu-es': 'Basque (Spain)',
    'fa-ir': 'Persian (Iran)',
    'fi-fi': 'Finnish (Finland)',
    'fil-ph': 'Filipino (Philippines)',
    'fr-ca': 'French (Canada)',
    'fr-fr': 'French (France)',
    'gl-es': 'Galician (Spain)',
    'gu-in': 'Gujarati (India)',
    'he-il': 'Hebrew (Israel)',
    'hi-in': 'Hindi (India)',
    'hr-hr': 'Croatian (Croatia)',
    'hu-hu': 'Hungarian (Hungary)',
    'hy-am': 'Armenian (Armenia)',
    'id-id': 'Indonesian (Indonesia)',
    'is-is': 'Icelandic (Iceland)',
    'it-it': 'Italian (Italy)',
    'ja-jp': 'Japanese (Japan)',
    'jv-id': 'Javanese (Indonesia)',
    'ka-ge': 'Georgian (Georgia)',
    'km-kh': 'Khmer (Cambodia)',
    'kn-in': 'Kannada (India)',
    'ko-kr': 'Korean (South Korea)',
    'lo-la': 'Lao (Laos)',
    'lt-lt': 'Lithuanian (Lithuania)',
    'lv-lv': 'Latvian (Latvia)',
    'ml-in': 'Malayalam (India)',
    'mr-in': 'Marathi (India)',
    'ms-my': 'Malay (Malaysia)',
    'nb-no': 'Norwegian Bokmal (Norway)',
    'ne-np': 'Nepali (Nepal)',
    'nl-nl': 'Dutch (Netherlands)',
    'pl-pl': 'Polish (Poland)',
    'pt-br': 'Portuguese (Brazil)',
    'pt-pt': 'Portuguese (Portugal)',
    'ro-ro': 'Romanian (Romania)',
    'ru-ru': 'Russian (Russia)',
    'si-lk': 'Sinhala (Sri Lanka)',
    'sk-sk': 'Slovak (Slovakia)',
    'sl-si': 'Slovenian (Slovenia)',
    'sr-rs': 'Serbian (Serbia)',
    'su-id': 'Sundanese (Indonesia)',
    'sv-se': 'Swedish (Sweden)',
    'sw-ke': 'Swahili (Kenya)',
    'sw-tz': 'Swahili (Tanzania)',
    'ta-in': 'Tamil (India)',
    'ta-lk': 'Tamil (Sri Lanka)',
    'ta-my': 'Tamil (Malaysia)',
    'ta-sg': 'Tamil (Singapore)',
    'te-in': 'Telugu (India)',
    'th-th': 'Thai (Thailand)',
    'tr-tr': 'Turkish (Turkey)',
    'uk-ua': 'Ukrainian (Ukraine)',
    'ur-in': 'Urdu (India)',
    'ur-pk': 'Urdu (Pakistan)',
    'vi-vn': 'Vietnamese (Vietnam)',
    'yue-hant-hk': 'Chinese, Cantonese (Traditional, Hong Kong)',
    'zu-za': 'Zulu (South Africa)'
}

OUTPUT_FORMAT = {
    'srt': 'SubRip',
    'ass': 'Advanced SubStation Alpha',
    'ssa': 'SubStation Alpha',
    'sub': 'MicroDVD Subtitle',
    'mpl2.txt': 'Similar to MicroDVD',
    'tmp': 'TMP Player Subtitle Format',
    'vtt': 'WebVTT',
    'json': 'json(Only times and text)',
    'ass.json': 'json(Complex ass content json)',
    'txt': 'Plain Text(Text or times)'
}

INPUT_FORMAT = {
    'vtt': 'WebVTT',
    'srt': 'SubRip',
    'ass': 'Advanced SubStation Alpha',
    'ssa': 'SubStation Alpha',
    'sub': 'MicroDVD Subtitle',
    'txt': 'Similar to MicroDVD(mpl2)',
    'tmp': 'TMP Player Subtitle Format',
    'json': 'json(Complex ass content json)'
}


def cmd_conversion(command):
    """
    Give a command and return a cross-platform command
    """
    if not IS_UNIX:
        cmd_args = command
    else:
        cmd_args = shlex.split(command)
    return cmd_args


def is_exe(file_path):
    """
    Checks whether a file is executable.
    """
    return os.path.isfile(file_path) and os.access(file_path, os.X_OK)


def which_exe(program_path):
    """
    Return the path for a given executable.
    """
    program_dir = os.path.split(program_path)[0]
    if program_dir:
        # if program directory exists
        if is_exe(program_path):
            return program_path
    else:
        # else find the program path
        for path in os.environ["PATH"].split(os.pathsep):
            path = path.strip('"')
            program_name = os.path.join(path, program_path)
            if is_exe(program_name):
                return program_name
        # if program located at app's directory
        program_name = os.path.join(APP_PATH, os.path.basename(program_path))
        if is_exe(program_name):
            return program_name
        # if program located at current directory
        program_name = os.path.join(os.getcwd(), os.path.basename(program_path))
        if is_exe(program_name):
            return program_name
    return None


def get_cmd(program_name):
    """
    Return the path for a given executable.
    "" returned when no executable exists.
    """
    if IS_UNIX:
        command = which_exe(program_name)
        if command:
            return command
    else:
        command = which_exe(program_name + ".exe")
        if command:
            return command

    return ""


def get_cmd_from_env(program_name, env_name):
    """
    Return the path or the value of environment variable
    for a given executable.
    """
    if env_name in os.environ:
        if is_exe(os.environ[env_name]):
            return os.environ[env_name]
        program_name = os.path.join(os.environ[env_name], program_name)
        if is_exe(program_name):
            return program_name
        return program_name + ".exe"
    return get_cmd(program_name)


FFMPEG_CMD = get_cmd_from_env("ffmpeg", "FFMPEG_PATH")
FFPROBE_CMD = get_cmd_from_env("ffprobe", "FFPROBE_PATH")
FFMPEG_NORMALIZE_CMD = get_cmd_from_env("ffmpeg-normalize", "FFMPEG_NORMALIZE_PATH")

DEFAULT_AUDIO_PRCS_CMDS = [
    FFMPEG_CMD + " -hide_banner -i \"{in_}\" -vn -af \"asplit[a],aphasemeter=video=0,\
ametadata=select:key=\
lavfi.aphasemeter.phase:value=-0.005:function=less,\
pan=1c|c0=c0,aresample=async=1:first_pts=0,[a]amix\" \
-ac 1 -f flac -loglevel error \"{out_}\"",
    FFMPEG_CMD + " -hide_banner -i \"{in_}\" -af \"lowpass=3000,highpass=200\" "
                 "-loglevel error \"{out_}\"",
    FFMPEG_NORMALIZE_CMD + " -v \"{in_}\" -ar 44100 -ofmt flac -c:a flac -pr -p -o \"{out_}\""
]

DEFAULT_AUDIO_CVT_CMD = \
    FFMPEG_CMD + " -hide_banner -y -i \"{in_}\" -vn -ac {channel} -ar {sample_rate}" \
                 " -loglevel error \"{out_}\""

DEFAULT_AUDIO_SPLT_CMD = \
    FFMPEG_CMD + " -y -ss {start} -i \"{in_}\" -t {dura} " \
    "-vn -ac [channel] -ar [sample_rate] -loglevel error \"{out_}\""

DEFAULT_VIDEO_FPS_CMD = FFPROBE_CMD + " -v 0 -of csv=p=0 -select_streams " \
                        "v:0 -show_entries stream=r_frame_rate \"{in_}\""

DEFAULT_CHECK_CMD = FFPROBE_CMD + " \"{in_}\" -show_format -pretty -loglevel quiet"

DEFAULT_ENGLISH_STOP_WORDS_SET_1 = \
    {'after', 'and', 'as', 'because', 'before', 'between', 'but', 'either', 'except', 'for', 'how',
     'include', 'including', 'includes', 'included', 'if', 'or', 'since', 'so', 'that', "that'll",
     'until', 'what', 'when', 'where', 'which', 'while', 'who', 'whom', 'why', 'with'}

DEFAULT_ENGLISH_STOP_WORDS_SET_2 = \
    {'a', 'about', 'above', 'after', 'again', 'against', 'ain', 'all', 'am', 'an', 'and', 'any',
     'are', 'aren', "aren't", 'as', 'at', 'be', 'because', 'been', 'before', 'being', 'below',
     'between', 'both', 'but', 'by', 'can', 'couldn', "couldn't", 'd', 'did', 'didn', "didn't",
     'do', 'does', 'doesn', "doesn't", 'doing', 'don', "don't", 'down', 'during', 'each', 'few',
     'for', 'from', 'further', 'had', 'hadn', "hadn't", 'has', 'hasn', "hasn't", 'have', 'haven',
     "haven't", 'having', 'he', 'her', 'here', 'hers', 'herself', 'him', 'himself', 'his', 'how',
     'i', 'if', 'in', 'into', 'is', 'isn', "isn't", 'it', "it's", 'its', 'itself', 'just', 'll',
     'm', 'ma', 'me', 'mightn', "mightn't", 'more', 'most', 'mustn', "mustn't", 'my', 'myself',
     'needn', "needn't", 'no', 'nor', 'not', 'now', 'o', 'of', 'off', 'on', 'once', 'only', 'or',
     'other', 'our', 'ours', 'ourselves', 'out', 'over', 'own', 're', 's', 'same', 'shan', "shan't",
     'she', "she's", 'should', "should've", 'shouldn', "shouldn't", 'so', 'some', 'such', 't',
     'than', 'that', "that'll", 'the', 'their', 'theirs', 'them', 'themselves', 'then', 'there',
     'these', 'they', 'this', 'those', 'through', 'to', 'too', 'under', 'until', 'up', 've', 'very',
     'was', 'wasn', "wasn't", 'we', 'were', 'weren', "weren't", 'what', 'when', 'where', 'which',
     'while', 'who', 'whom', 'why', 'will', 'with', 'won', "won't", 'wouldn', "wouldn't", 'y',
     'you', "you'd", "you'll", "you're", "you've", 'your', 'yours', 'yourself', 'yourselves'
     }
# Reference: https://gist.github.com/sebleier/554280
